xrmocap/ops/bottom_up_association/associate/fourdag_associate.py

```diff
@@ -0,0 +1,956 @@
+# yapf: disable
+import copy
+import heapq
+import logging
+import numpy as np
+from typing import Union
+
+from xrmocap.utils.fourdag_utils import (
+    LIMB_INFO, line2linedist, point2linedist, welsch,
+)
+
+# yapf: enable
+
+
+class Clique():
+
+    def __init__(self, paf_id, proposal, score=-1) -> None:
+        """class for limb clique, which is used for solve 4D graph paf_id
+        (int):
+
+        paf index proposal (List):     a list of allocated bone index to the
+        clique score (float):     the score of the clique, larger score will be
+        solve earlier
+        """
+        self.paf_id = paf_id
+        self.proposal = proposal
+        self.score = score
+
+    def __lt__(self, other):
+        if self.score > other.score:
+            return True
+        else:
+            return False
+
+
+class Voting():
+
+    def __init__(self) -> None:
+        """vote class for clique it will record the kps haven been allocated
+        and it will be used to solve graph."""
+        self.fst = np.zeros(2, dtype=np.int8)
+        self.sec = np.zeros(2, dtype=np.int8)
+        self.fst_cnt = np.zeros(2, dtype=np.int8)
+        self.sec_cnt = np.zeros(2, dtype=np.int8)
+        self.vote = dict()
+
+    def parse(self):
+        self.fst_cnt = np.zeros(2)
+        self.sec_cnt = np.zeros(2)
+        if len(self.vote) == 0:
+            return
+
+        _vote = copy.deepcopy(self.vote)
+        for i in range(2):
+            for index in range(2):
+                person_id = max(_vote, key=lambda x: _vote[x][index])
+
+                if i == 0:
+                    self.fst[index] = person_id
+                    self.fst_cnt[index] = _vote[person_id][index]
+                else:
+                    self.sec[index] = person_id
+                    self.sec_cnt[index] = _vote[person_id][index]
+                _vote[person_id][index] = 0
+
+
+class Camera():
+
+    def __init__(self, cam_param) -> None:
+        super().__init__()
+        c_K = cam_param.intrinsic33()
+        c_T = np.array(cam_param.get_extrinsic_t())
+        c_R = np.array(cam_param.get_extrinsic_r())
+        c_Ki = np.linalg.inv(c_K)
+        self.c_Rt_Ki = np.matmul(c_R.T, c_Ki)
+        self.Pos = -np.matmul(c_R.T, c_T)
+
+    def cal_ray(self, uv):
+        var = -self.c_Rt_Ki.dot(np.append(uv, 1).T)
+        return var / np.linalg.norm(var)
```

Comment:
@LazyBusyYang Is it necessary for primer to support such operation?

Reply:
Negative. Haven't seen another use of this function.